File: LearningLinearAlgebra.FSharp.Tests/CartesianRepresentationTests.fs
namespace LearningLinearAlgebra.Numbers

open Xunit

module CartesianRepresentationTests =

    open LearningLinearAlgebra
    open Complex

    [<Fact>]
    let ``Sum of two complex numbers is calculated as sum of the components`` () =
        let a = C(5, 7)
        let b = C(11, 13)

        let sum = Add a b

        sum |> Should.BeEquivalentTo.Complex.Number(C(16, 20))
        (a + b) |> Should.BeEquivalentTo.Complex.Number(sum)

    [<Fact>]
    let ``Sum of complex numbers is commutative`` () =
        let a = C(5, 7)
        let b = C(11, 13)

        let aPlusB = Add a b
        let bPlusA = Add b a

        bPlusA |> Should.BeEquivalentTo.Complex.Number(aPlusB)

    [<Fact>]
    let ``Sum of complex numbers is associative`` () =
        let a = C(5, 7)
        let b = C(11, 13)
        let c = C(17, 19)

        let sumOfAndBPlusC = Add (Add a b) c
        let aPlusSumOfBAndC = Add a (Add b c)

        aPlusSumOfBAndC |> Should.BeEquivalentTo.Complex.Number(sumOfAndBPlusC)

    [<Fact>]
    let ``Difference of two complex numbers is calculated as difference of the components`` () =
        let a = C(5, 7)
        let b = C(11, 13)

        let difference = Subtract a b

        difference |> Should.BeEquivalentTo.Complex.Number(C(-6.0, -6.0))
        a - b |> Should.BeEquivalentTo.Complex.Number(Subtract a b)

    [<Fact>]
    let ``Difference of complex numbers is not commutative`` () =
        let a = C(5, 7)
        let b = C(11, 13)

        let aPlusB = Subtract a b
        let bPlusA = Subtract b a

        Assert.NotEqual(aPlusB, bPlusA)

    [<Fact>]
    let ``Difference of complex numbers is not associative`` () =
        let a = C(5, 7)
        let b = C(11, 13)
        let c = C(17, 19)

        let differenceOfAndBPlusC = Subtract (Subtract a b) c

        let aPlusDifferenceOfBAndC = Subtract a (Subtract b c)

        Assert.NotEqual(differenceOfAndBPlusC, aPlusDifferenceOfBAndC)

    [<Fact>]
    let ``Product of complex numbers (a1, b1) and (a2, b2) is calculated as (a1*a2 - b1*b2, a1*b2 + a2*b1)`` () =
        let a = C(5, 7)
        let b = C(11, 13)

        let product = Multiply a b

        product |> Should.BeEquivalentTo.Complex.Number(C(-36.0, 142.0))
        a * b |> Should.BeEquivalentTo.Complex.Number(Multiply a b)

    [<Fact>]
    let ``Product of complex numbers is commutative`` () =
        let a = C(5, 7)
        let b = C(11, 13)

        let aMultipliedByB = Multiply a b
        let bMultipliedByA = Multiply b a

        bMultipliedByA |> Should.BeEquivalentTo.Complex.Number(aMultipliedByB)

    [<Fact>]
    let ``Product of complex numbers is associative`` () =
        let a = C(5, 7)
        let b = C(11, 13)
        let c = C(17, 19)

        let productOfAndBMultipliedByC = Multiply (Multiply a b) c

        let aMultipliedByProductOfBAndC = Multiply a (Multiply b c)

        aMultipliedByProductOfBAndC
        |> Should.BeEquivalentTo.Complex.Number(productOfAndBMultipliedByC)

    [<Fact>]
    let ``Multiplication of complex numbers distributes over addition`` () =
        let a = C(5, 7)
        let b = C(11, 13)
        let c = C(17, 19)

        let left = Multiply a (Add b c)

        let right = Add (Multiply a b) (Multiply a c)

        right |> Should.BeEquivalentTo.Complex.Number(left)

    [<Fact>]
    let ``Square of complex number is the number multiplied by itself`` () =
        let complex = C(5, 7)

        let squared = Square complex

        squared |> Should.BeEquivalentTo.Complex.Number(complex * complex)

    [<Fact>]
    let ``Quotient of complex numbers is calculated calculated using complex conjucations`` () =
        let a = C(-2, 1)
        let b = C(1, 2)

        let quotient = Divide a b

        quotient |> Should.BeEquivalentTo.Complex.Number(C(0.0, 1.0))

    [<Fact>]
    let ``Another example of division of complex numbers`` () =
        let a = C(0, 3)
        let b = C(-1, -1)

        let quotient = Divide a b
        let f = -3.0 / 2.0

        quotient |> Should.BeEquivalentTo.Complex.Number(C(-3.0 / 2.0, -3.0 / 2.0))

    [<Fact>]
    let ``Quotient of complex numbers is not commutative`` () =
        let a = C(5, 7)
        let b = C(11, 13)

        let aDividedByB = Divide a b
        let bDividedByA = Divide b a

        Assert.NotEqual(aDividedByB, bDividedByA)

    [<Fact>]
    let ``Quotient of complex numbers is not associative`` () =
        let a = C(5, 7)
        let b = C(11, 13)
        let c = C(17, 19)

        let quotientOfAndBDividedByC = Divide (Divide a b) c

        let aDividedByQuotientOfBAndC = Divide a (Divide b c)

        Assert.NotEqual(quotientOfAndBDividedByC, aDividedByQuotientOfBAndC)

    [<Fact>]
    let ``Modulus of complex number is the square root of sum of second power of components`` () =
        let a = C(5, 7)

        let modulus = Modulus a

        modulus |> Should.BeEquivalentTo.Real.Number(sqrt (74.0))

    [<Fact>]
    let ``Another example of modulus`` () =
        let a = C(1, -1)

        let modulus = Modulus a

        modulus |> Should.BeEquivalentTo.Real.Number(sqrt (2.0))

    [<Fact>]
    let ``Sum of complex and it's addivive inverse is zero`` () =
        let a = C(2, -3)

        a + (-a) |> Should.BeEquivalentTo.Complex.Number(C(0, 0))

File: LearningLinearAlgebra.Tests/Numbers/CartesianRepresentationTests.cs
using Computation.Numbers;
using FluentAssertions;
using Xunit;
using static Computation.Numbers.ComplexNumber<float>;

namespace LearningLinearAlgebra.Tests.Numbers;

public class CartesianRepresentationTests
{
    [Fact]
    public void Sum_of_two_complex_numbers_is_calculated_as_sum_of_components()
    {
        var a = C(5, 7);
        var b = C(11, 13);

        var sum = Add(a, b);

        sum.Should().Be(C(16, 20));
        (a + b).Should().Be(Add(a, b));
        a.Add(b).Should().Be(Add(a, b));
    }

    [Fact]
    public void Sum_of_complex_numbers_is_commutative()
    {
        var a = C(5, 7);
        var b = C(11, 13);

        var aPlusB = Add(a, b);
        var bPlusA = Add(b, a);

        aPlusB.Should().Be(bPlusA);
    }

    [Fact]
    public void Sum_of_complex_numbers_is_associative()
    {
        var a = C(5, 7);
        var b = C(11, 13);
        var c = C(17, 19);

        var sumOfAndBPlusC = Add(Add(a, b), c);
        var aPlusSumOfBAndC = Add(a, Add(b, c));

        sumOfAndBPlusC.Should().Be(aPlusSumOfBAndC);
    }

    [Fact]
    public void Difference_of_two_complex_numbers_is_calculated_as_difference_of_components()
    {
        var a = C(5, 7);
        var b = C(11, 13);

        var difference = Subtract(a, b);

        difference.Should().Be(C(-6, -6));
        (a - b).Should().Be(Subtract(a, b));
        a.Subtract(b).Should().Be(Subtract(a, b));
    }

    [Fact]
    public void Difference_of_complex_numbers_is_not_commutative()
    {
        var a = C(5, 7);
        var b = C(11, 13);

        var aMinusB = Subtract(a, b);
        var bMinusA = Subtract(b, a);

        aMinusB.Should().NotBe(bMinusA);
    }

    [Fact]
    public void Difference_of_complex_numbers_is_not_associative()
    {
        var a = C(5, 7);
        var b = C(11, 13);
        var c = C(17, 19);

        var differenceOfAndBMinusC = Subtract(Subtract(a, b), c);
        var aMinusDifferenceOfBAndC = Subtract(a, Subtract(b, c));

        differenceOfAndBMinusC.Should().NotBe(aMinusDifferenceOfBAndC);
    }

    [Fact]
    public void Product_of_complex_numbers()
    // Product of complex numbers (a1, b1) and (a2, b2) is calculated as (a1*a2 - b1*b2, a1*b2 + a2*b1)
    {
        var a = C(5, 7);
        var b = C(11, 13);

        var product = Multiply(a, b);

        product.Should().Be(C(-36, 142));
        (a * b).Should().Be(Multiply(a, b));
        a.Multiply(b).Should().Be(Multiply(a, b));
    }

    [Fact]
    public void Product_of_complex_numbers_is_commutative()
    {
        var a = C(5, 7);
        var b = C(11, 13);

        var aMultipliedByB = Multiply(a, b);
        var bMultipliedByA = Multiply(b, a);

        aMultipliedByB.Should().Be(bMultipliedByA);
    }

    [Fact]
    public void Product_of_complex_numbers_is_associative()
    {
        var a = C(5, 7);
        var b = C(11, 13);
        var c = C(17, 19);

        var productOfAndBMultipliedByC = Multiply(Multiply(a, b), c);
        var aMultipliedByProductOfBAndC = Multiply(a, Multiply(b, c));

        productOfAndBMultipliedByC.Should().Be(aMultipliedByProductOfBAndC);
    }

    [Fact]
    public void Multiplication_of_complex_numbers_distributes_over_addition()
    {
        var a = C(5, 7);
        var b = C(11, 13);
        var c = C(17, 19);

        var left = Multiply(a, Add(b, c));
        var right = Add(Multiply(a, b), Multiply(a, c));

        left.Should().Be(right);
    }

    [Fact]
    public void Product_of_complex_number_and_real_number_is_calculated_by_multiplying_both_components_of_complex_by_real()
    {
        var a = 5.0f;
        var b = C(-11, 5);

        var product = Multiply(a, b);

        product.Should().Be(C(-55, 25));
        (a * b).Should().Be(Multiply(a, b));
        (b * a).Should().Be(Multiply(a, b));
        a.Multiply(b).Should().Be(Multiply(a, b));
        b.Multiply(a).Should().Be(Multiply(a, b));
    }

    [Fact]
    public void Quotient_of_complex_numbers_is_calculated_using_complex_conjugations()
    {
        var a = C(-2, 1);
        var b = C(1, 2);

        var quotient = Divide(a, b);

        quotient.Should().Be(C(0, 1));
        a.Divide(b).Should().Be(Divide(a, b));
    }

    [Fact]
    public void Another_example_of_division_of_complex_numbers()
    {
        var a = C(0, 3);
        var b = C(-1, -1);

        var quotient = Divide(a, b);

        quotient.Should().Be(C(-3.0f / 2.0f, -3.0f / 2.0f));
    }

    [Fact]
    public void Quotient_of_complex_numbers_is_not_commutative()
    {
        var a = C(5, 7);
        var b = C(11, 13);

        var aDividedByB = Divide(a, b);
        var bDividedByA = Divide(b, a);

        aDividedByB.Should().NotBe(bDividedByA);
    }

    [Fact]
    public void Quotient_of_complex_numbers_is_not_associative()
    {
        var a = C(5, 7);
        var b = C(11, 13);
        var c = C(17, 19);

        var quotientOfAndBDividedByC = Divide(Divide(a, b), c);
        var aDividedByQuotientOfBAndC = Divide(a, Divide(b, c));

        quotientOfAndBDividedByC.Should().NotBe(aDividedByQuotientOfBAndC);
    }

    [Fact]
    public void Square_of_complex_number_is_number_multiplied_by_itself()
    {
        var complex = C(5, 7);

        var squared = Square(complex);

        squared.Should().Be(complex * complex);
        complex.Square().Should().Be(Square(complex));
    }

    [Fact]
    public void Modulus_of_complex_number_is_square_root_of_sum_of_second_power_of_components()
    {
        var complex = C(5, 7);

        var modulus = Modulus(complex);

        modulus.Should().Be(float.Sqrt(74.0f));
        complex.Modulus().Should().Be(Modulus(complex));
    }

    [Fact]
    public void Another_example_of_modulus()
    {
        var complex = C(1, -1);

        var modulus = Modulus(complex);

        modulus.Should().Be(float.Sqrt(2));
    }

    [Fact]
    public void Sum_of_complex_and_its_additive_inverse_is_zero()
    {
        var a = C(2, -3);

        (a + -a).Should().Be(C(0, 0));
    }

    [Fact]
    public void Complex_can_be_rounded_to_zero_if_it_differs_from_zero_only_by_small_value()
    {
        var almostZero = C(0.0000001f, 0.0000001f);

        Round(almostZero).Should().Be(Zero);
        almostZero.Round().Should().Be(Zero);
    }

    [Fact]
    public void Complex_cannot_be_rounded_to_zero_if_it_differs_from_zero_more_than_small_value()
    {
        var almostZero = C(0.000001f, 0.000001f);

        Round(almostZero).Should().NotBe(Zero);
        almostZero.Round().Should().NotBe(Zero);
    }

    [Fact]
    public void Complex_can_be_rounded_to_one_if_it_differs_from_one_only_by_small_value()
    {
        var almostOne = C(0.9999999f, 0.0000001f);

        Round(almostOne).Should().Be(One);
        almostOne.Round().Should().Be(One);
    }

    [Fact]
    public void Complex_cannot_be_rounded_to_as_one_if_it_differs_from_one_more_than_small_value()
    {
        var almostOne = C(0.999999f, 0.000001f);

        Round(almostOne).Should().NotBe(One);
        almostOne.Round().Should().NotBe(One);
    }
}
